sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala

```diff
@@ -0,0 +1,209 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.command
+
+import scala.collection.mutable
+
+import org.apache.spark.sql._
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.EliminateSubqueryAliases
+import org.apache.spark.sql.catalyst.catalog.{CatalogRelation, CatalogTable}
+import org.apache.spark.sql.catalyst.plans.logical.{BasicColStats, Statistics}
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+
+/**
+ * Analyzes the given columns of the given table in the current database to generate statistics,
+ * which will be used in query optimizations.
+ */
+case class AnalyzeColumnCommand(
+    tableName: String,
+    columnNames: Seq[String]) extends RunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val sessionState = sparkSession.sessionState
+    val tableIdent = sessionState.sqlParser.parseTableIdentifier(tableName)
+    val relation = EliminateSubqueryAliases(sessionState.catalog.lookupRelation(tableIdent))
+
+    // check correctness for column names
+    val attributeNames = relation.output.map(_.name.toLowerCase)
+    val invalidColumns = columnNames.filterNot { col => attributeNames.contains(col.toLowerCase)}
+    if (invalidColumns.nonEmpty) {
+      throw new AnalysisException(s"Invalid columns for table $tableName: $invalidColumns.")
+    }
+
+    relation match {
+      case catalogRel: CatalogRelation =>
+        updateStats(catalogRel.catalogTable,
+          AnalyzeTableCommand.calculateTotalSize(sparkSession, catalogRel.catalogTable))
+
+      case logicalRel: LogicalRelation if logicalRel.catalogTable.isDefined =>
+        updateStats(logicalRel.catalogTable.get, logicalRel.relation.sizeInBytes)
+
+      case otherRelation =>
+        throw new AnalysisException(s"ANALYZE TABLE is not supported for " +
+          s"${otherRelation.nodeName}.")
+    }
+
+    def updateStats(catalogTable: CatalogTable, newTotalSize: Long): Unit = {
+      val lowerCaseNames = columnNames.map(_.toLowerCase)
+      val attributes =
+        relation.output.filter(attr => lowerCaseNames.contains(attr.name.toLowerCase))
+
+      // collect column statistics
+      val aggColumns = mutable.ArrayBuffer[Column](count(Column("*")))
+      attributes.foreach(entry => aggColumns ++= statsAgg(entry.name, entry.dataType))
+      val statsRow: InternalRow = Dataset.ofRows(sparkSession, relation).select(aggColumns: _*)
+        .queryExecution.toRdd.collect().head
+
+      // We also update table-level stats to prevent inconsistency in case of table modification
```

Comment:
If we can trust the `nullable` of attributes, can we skip calculating the number of null for the attribute which is not nullable.